server/src/com/thoughtworks/go/server/materials/ConfigMaterialUpdater.java

```diff
@@ -0,0 +1,135 @@
+                                 MaterialChecker materialChecker,
+                                 ConfigMaterialUpdateCompletedTopic configCompletedTopic,
+                                 MaterialUpdateCompletedTopic topic,
+                                 MaterialService materialService,
+                                 SubprocessExecutionContext subprocessExecutionContext)
+    {
+        this.repoConfigDataSource = repoConfigDataSource;
+        this.materialChecker = materialChecker;
+        this.materialRepository = materialRepository;
+        this.configCompleted = configCompletedTopic;
+        this.topic = topic;
+        this.materialService = materialService;
+        this.subprocessExecutionContext = subprocessExecutionContext;
+
+        this.configCompleted.addListener(this);
+    }
+
+    @Override
+    public void onMessage(MaterialUpdateCompletedMessage message) {
+        Material material = message.getMaterial();
+        //MDU is done using the checkout, it has done db update and stored latest changes
+        // but MUS is still waiting for material updated message on MaterialUpdateCompletedTopic
+        if (LOGGER.isDebugEnabled()) {
+            LOGGER.debug(format("[Config Material Update] Config material update completed for material %s. Starting parse process", material));
+        }
+        try {
+            if(message instanceof MaterialUpdateFailedMessage)
+            {
+                MaterialUpdateFailedMessage failure = (MaterialUpdateFailedMessage)message;
+                LOGGER.warn(String.format("[Config Material Update] Cannot update configuration part because material update has failed. Reason: ",
+                        failure.getReason()));
+            }
+            else {
+                File folder = materialRepository.folderFor(material);
+                MaterialRevisions latestModification = materialRepository.findLatestModification(material);
+                Revision revision = latestModification.firstModifiedMaterialRevision().getRevision();
+
+                MaterialRevision lastParseRevision = getMaterialRevisionAtLastParseAttempt(message);
+                if (lastParseRevision == null) {
+                    //never parsed
+                    UpdateConfigurationFromCheckout(folder, revision, material);
+                } else if (latestModification.findRevisionFor(material.config())
+                        .hasChangedSince(lastParseRevision)) {
+                    // revision has changed. the config files might have been updated
+                    UpdateConfigurationFromCheckout(folder, revision, material);
+                } else {
+                    // revision is the same as last time, no need to parse again
+                }
+            }
+        }
+        finally {
+            // always post the original message further
+            // this will remove material from inProgress in MUS
+            topic.post(message);
+        }
+    }
+
```

Comment:
Does it make sense to spend effort specifically for P4 and TFS, given that they're more cumbersome to manage and can potentially cause problems if we don't get the workspaces right? If it's little effort (and it might be), then, it's ok.